peps/pep-0705.rst

```diff
@@ -118,175 +116,84 @@ However, this no longer works once you start nesting dictionaries::
+            return f'{movie["name"]} ({movie["year"]})'
 
-    from typing import NotRequired, TypedDict
+A mixture of read-only and non-read-only items is permitted, allowing the second motivating example to be correctly annotated::
 
-    class Movie(TypedDict, readonly=True):
-        name: str
-        director: str
+    class HasTimestamp(TypedDict):
+        timestamp: float
 
-    class Book(TypedDict, readonly=True):
-        name: str
-        author: str
+    class HasTimestampedMetadata(TypedDict):
+        metadata: ReadOnly[HasTimestamp]
 
-In addition to these benefits, by flagging arguments of a function as read-only (by using a read-only ``TypedDict`` like ``Movie`` or ``Book``), it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desireable property of a function interface.
+    def update_metadata_timestamp(d: HasTimestampedMetadata) -> None:
+        d["metadata"]["timestamp"] = now()
 
-A new ``typing.ReadOnly`` type qualifier allows removing the ability to mutate individual entries, permitting a mixture of readonly and mutable entries. This is necessary for supporting the second motivating example, updating nested dicts::
+    class Logs(HasTimestamp):
+        loglines: list[str]
 
     class UserAudit(TypedDict):
         name: str
-        metadata: ReadOnly[Logs]
-
-This PEP only proposes making ``ReadOnly`` valid in a ``TypedDict``. A possible future extension would be to support it in additional contexts, such as in protocols.
-
-Finally, to support type discrimination, we add a second flag to ``TypedDict``, ``other_keys``, which when set to ``typing.Never``, prevents instances from holding any key not explicitly listed in the type::
-
-    class EntertainmentMovie(TypedDict, readonly=True, other_keys=Never):
-        movie: Movie
+        metadata: Logs
 
-    class EntertainmentBook(TypedDict, readonly=True, other_keys=Never):
-        book: Book
+    def rename_user(d: UserAudit, name: str) -> None:
+        d["name"] = name
+        update_metadata_timestamp(d)  # Now OK
 
-    Entertainment = EntertainmentMovie | EntertainmentBook
+In addition to these benefits, by flagging arguments of a function as read-only (by using a ``TypedDict`` like ``Movie`` with read-only items), it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
 
-    def get_name(entertainment: Entertainment) -> str:
-        if "movie" in entertainment:
-            return entertainment["movie"]["name"]
-        elif "book" in entertainment:
-            return entertainment["book"]["name"]
-        else:
-            raise ValueError("Unexpected entertainment type")
+This PEP only proposes making ``ReadOnly`` valid in a ``TypedDict``. A possible future extension would be to support it in additional contexts, such as in protocols.
```

Comment:
Grammar nit: Move the "only" to later in the sentence: "This PEP proposes making ``ReadOnly`` valid only in a ``TypedDict``."